remote-vfs.c

```diff
@@ -0,0 +1,1889 @@
+
+	lookup_main_url();
+
+	if (setup_cache_server)
+		select_cache_server();
+
+	select_odb();
+}
+
+/*
+ * Request gvfs/config from main Git server.  Print the received
+ * server configuration.
+ */
+static int do_sub_cmd__config(int argc, const char **argv)
+{
+	struct strbuf config_data = STRBUF_INIT;
+	int ret = HTTP_OK;
+
+	trace2_cmd_mode("config");
+
+	finish_init(0);
+
+	trace2_region_enter("remote-vfs", "config", NULL);
+	ret = rest__get__gvfs_config(&config_data);
+	trace2_region_leave("remote-vfs", "config", NULL);
+	
+	if (ret != HTTP_OK) {
+		// TODO print error
+	}
+	else
+		printf("%s\n", config_data.buf);
+
+	strbuf_release(&config_data);
+
+	return ret;
+}
+
+/*
+ * Bulk fetch a list of missing objects.  Read the list from stdin
+ * assuming it comes from `rev-list` and has '?' markers.
+ *
+ * Create a packfile (or a loose object if only one OID requested).
+ */
+static int do_sub_cmd__get_missing(int argc, const char **argv)
+{
+	static int depth = 1;
+
+	static struct option get_missing_options[] = {
+		OPT_INTEGER('d', "depth", &depth,
+			    N_("Commit depth")),
+		OPT_CALLBACK(0, "cache-server", NULL,
+			     N_("cache-server"),
+			     N_("cache-server=disable|allow|verify|error"),
+			     option_parse_csu),
+		OPT_END(),
+	};
+
+	struct oidset oids = OIDSET_INIT;
+	struct oidset_iter iter;
+	struct strbuf output_pathname = STRBUF_INIT;
+	int count;
+	int ret = HTTP_OK;
+
+	trace2_cmd_mode("get-missing");
+
+	if (argc > 1 && !strcmp(argv[1], "-h"))
+		usage_with_options(get_missing_usage, get_missing_options);
+
+	argc = parse_options(argc, argv, NULL, get_missing_options,
+			     get_missing_usage, 0);
+
+	finish_init(1);
+
+	count = read_stdin_from_rev_list(&oids, 1);
+	trace2_data_intmax("remote-vfs", NULL, "get-missing/count", count);
+
+	switch (count) {
+	case 0:
+		printf("emptyset\n");
+		ret = HTTP_OK;
+		break;
+
+	case 1:
+		oidset_iter_init(&oids, &iter);
+		do__loose__gvfs_object(oidset_iter_next(&iter),
+				       &output_pathname);
+		printf("loose %s\n", output_pathname.buf);
+		break;
+
+	default:
+		do__packfile__gvfs_objects(&oids, depth, &output_pathname);
+
+		printf("packfile %d %s\n", count, output_pathname.buf);
+		break;
+	}
+
+	strbuf_release(&output_pathname);
+	oidset_clear(&oids);
+	return ret;
+}
+
+static int do_sub_cmd(int argc, const char **argv)
+{
+	if (!strcmp(argv[0], "get-missing"))
+		return do_sub_cmd__get_missing(argc, argv);
+
+	if (!strcmp(argv[0], "config"))
+		return do_sub_cmd__config(argc, argv);
+
+	// TODO have "interactive" mode that behaves like
+	// TODO current read-object-hook and could do a
+	// TODO series of batches or individual items.
+	// TODO
+	// TODO have "test" mode that could be used to drive
+	// TODO unit testing.
+
+	return RV_ERROR__USAGE;
+}
+
+/*
+ * Connect to primary Git server or a cache-server using the GVFS Protocol.
+ * https://github.com/microsoft/VFSForGit/blob/master/Protocol.md
+ */
+int cmd_main(int argc, const char **argv)
```

Comment:
Is this command _actually_ intended to be called transparently when the user specifies the `vfs://` protocol? If not, it should not be named `git-remote-<protocol>` but should be renamed instead (something like `git-vfs-helper` or `git-vfs-sync`).

Reply:
ok. was wondering about that.  naming is hard.  :-)